spec.html

```diff
@@ -36262,8 +36282,17 @@ <h1>
               1. Perform ! CreateDataPropertyOrThrow(_A_, ! ToString(𝔽(_i_)), _capturedValue_).
               1. If the _i_<sup>th</sup> capture of _R_ was defined with a |GroupName|, then
                 1. Let _s_ be the CapturingGroupName of that |GroupName|.
-                1. Perform ! CreateDataPropertyOrThrow(_groups_, _s_, _capturedValue_).
-                1. Append _s_ to _groupNames_.
+                1. Let _isMatchedEarlier_ be *false*.
+                1. For each integer _j_ such that _j_ &ge; 1 and _j_ &lt; _i_, do
```

Comment:
Another possible simplification:
 ```patch
             1. Perform ! CreateDataPropertyOrThrow(_A_, *"groups"*, _groups_).
+            1. Let _matchedGroupNames_ be a new empty List.
             1. For each integer _i_ such that _i_ &ge; 1 and _i_ &le; _n_, in ascending order, do
               1. Let _captureI_ be _i_<sup>th</sup> element of _r_'s _captures_ List.
               …
               1. If the _i_<sup>th</sup> capture of _R_ was defined with a |GroupName|, then
                 1. Let _s_ be the CapturingGroupName of that |GroupName|.
-                1. Let _isMatchedEarlier_ be *false*.
-                1. For each integer _j_ such that _j_ &ge; 1 and _j_ &lt; _i_, do
-                  1. If the _j_<sup>th</sup> capture of _R_ was defined with a |GroupName|, then
-                    1. Let _sj_ be the CapturingGroupName of that |GroupName|.
-                    1. If SameValue(_sj_, _s_) is *true* and the _j_<sup>th</sup> element of _r_'s _captures_ List is not *undefined*, set _isMatchedEarlier_ to *true*.
-                1. If _isMatchedEarlier_ is *false*, then
+                1. If _matchedGroupNames_ contains _s_, then
+                  1. Set _s_ to *undefined*.
+                1. Else,
+                  1. If _capturedValue_ is not *undefined*, append _s_ to _matchedGroupNames_.
                   1. NOTE: If there are multiple groups named _s_, _groups_ may already have an _s_ property at this point. However, because _groups_ is an ordinary object whose properties are all writable data properties, the call to CreateDataPropertyOrThrow is nevertheless guaranteed to succeed.
                   1. Perform ! CreateDataPropertyOrThrow(_groups_, _s_, _capturedValue_).
-                  1. Append _s_ to _groupNames_.
-                1. Else,
-                  1. Append *undefined* to _groupNames_.
+                1. Append _s_ to _groupNames_.
               1. Else,
                 1. Append *undefined* to _groupNames_.
 ```

Reply:
Oh, nice. Pushed a [slight variation](https://github.com/tc39/ecma262/commit/61e7436a9e2dafdc79b1cc24f4504e4c52ef8bb3) which doesn't involve mutating _s_.